components/api_server/src/model/actions.py

```diff
@@ -7,73 +7,66 @@
 from shared.model.metric import Metric
 from shared.model.source import Source
 from shared.utils.type import ItemId
-from shared_data_model import DATA_MODEL
 
 from model.report import Report
 from utils.functions import uuid
 from utils.type import Position
 
 
-def copy_item(item, **kwargs):
+def copy_item[Item: Metric | Report | Source | Subject](item: Item, **kwargs) -> Item:
     """Return a copy of the item."""
-    item_copy = item.copy()
+    item_copy = cast(Item, item.copy())
     for key, value in kwargs.items():
         item_copy[key] = value
     return item_copy
 
 
-def copy_source(source, change_name: bool = True):
+def copy_source(source: Source) -> Source:
     """Return a copy of the source."""
-    kwargs = {}
-    if change_name:
-        kwargs["name"] = f"{source.get('name') or DATA_MODEL.sources[source['type']].name} (copy)"
-    return copy_item(source, **kwargs)
+    return copy_item(source)
 
 
-def copy_metric(metric, change_name: bool = True):
+def copy_metric(metric: Metric) -> Metric:
     """Return a copy of the metric and its sources."""
     kwargs: dict[str, Any] = {
-        "sources": {uuid(): copy_source(source, change_name=False) for source in metric["sources"].values()},
+        "sources": {uuid(): copy_source(source) for source in metric["sources"].values()},
```

Comment:
Why does the `Metric.copy()` not implement this instead?

Reply:
Because the Metric class is (still) a dict subclass, so `Metric.copy()` is `dict.copy()`. I think we should first convert `Metric` (and `Subject`, `Source`, etc.) to a proper class before moving these methods there.